src/libopensc/card-piv.c

```diff
@@ -467,208 +795,1885 @@ static struct sc_card_driver piv_drv = {
+
+	const char *option = NULL;
+	int piv_max_object_size_found = 0;
+#ifdef ENABLE_PIV_SM
+	int piv_pairing_code_found = 0;
+	int piv_use_sm_found = 0;
+#endif
+
+	if ((option = getenv("PIV_MAX_OBJECT_SIZE")) != NULL) {
+		sc_log(card->ctx, "getenv(\"PIV_MAX_OBJECT_SIZE\")=\"%s\"", option);
+		priv->max_object_size = atoi(option);
+		if (priv->max_object_size < PIV_MAX_OBJECT_SIZE || priv->max_object_size > MAX_FILE_SIZE) {
+			sc_log(card->ctx,"Invalid max_object_size: \"%d\"", priv->max_object_size);
+			if (priv->max_object_size < PIV_MAX_OBJECT_SIZE)
+				priv->max_object_size  = PIV_MAX_OBJECT_SIZE;
+			else 
+				priv->max_object_size = MAX_FILE_SIZE; /* conserative value if error */
+		} else 
+			piv_max_object_size_found = 1;
+		sc_log(card->ctx," priv->max_object_size:%d", priv->max_object_size);
+	}
+
+#ifdef ENABLE_PIV_SM
+	/* pairing code is 8 decimal digits and is card specific */
+	if ((option = getenv("PIV_PAIRING_CODE")) != NULL) {
+		sc_log(card->ctx,"getenv(\"PIV_PAIRING_CODE\")=\"%s\"", option);
+		memcpy(priv->pairing_code, option, MIN(strlen(option), 8));
+		piv_pairing_code_found = 1;
+	}
+
+	if ((option = getenv("PIV_USE_SM"))!= NULL) {
+		sc_log(card->ctx,"getenv(\"PIV_USE_SM\")=\"%s\"", option);
+		if (!strcmp(option, "never")) {
+			priv->sm_flags |= PIV_SM_FLAGS_NEVER;
+			piv_use_sm_found = 1;
+		}
+		else if (!strcmp(option, "always")) {
+			priv->sm_flags |= PIV_SM_FLAGS_ALWAYS;
+			piv_use_sm_found = 1;
+		}
+		else {
+			sc_log(card->ctx,"Invalid piv_use_sm: \"%s\"", option);
+		}
+	}
+#endif
+
+	for (i = 0; card->ctx->conf_blocks[i]; i++) {
+		found_blocks = scconf_find_blocks(card->ctx->conf, card->ctx->conf_blocks[i],
+				"card_driver", "PIV-II");
+		if (!found_blocks)
+			continue;
+
+		for (j = 0, block = found_blocks[j]; block; j++, block = found_blocks[j]) {
+			
+#ifdef ENABLE_PIV_SM
+			/*
+			 * "piv_use_sm" if card supports NIST sp800-73-4 sm, when should it be used 
+			 * never - use card like 800-73-3, i.e. contactless is very limited on 
+			 * true PIV cards. Some  PIV-like" card may allow this. 
+			 * this security risk
+			 * always - Use even for contact interface.
+			 * PINS, crypto and reading of object will not show up in logs
+			 * or over network. 
+			 */
+
+			if (piv_use_sm_found == 0) {
+				option = scconf_get_str(block, "piv_use_sm", "default");
+				sc_log(card->ctx,"conf: \"piv_use_sm\"=\"%s\"", option);
+				if  (strcmp(option,"default")) {
+					if (!strcmp(option, "never")) {
+						priv->sm_flags |= PIV_SM_FLAGS_NEVER;
+					}
+					else if (!strcmp(option, "always")) {
+						priv->sm_flags |= PIV_SM_FLAGS_ALWAYS;
+					}
+					else {
+						sc_log(card->ctx,"Invalid piv_use_sm: \"%s\"", option);
+					}
+				}
+			}
+
+			/* This is really a card specific value and should not be in the conf file */
+			if (piv_pairing_code_found == 0) {
+				option = scconf_get_str(block, "piv_pairing_code", "00000000");
+				memcpy(priv->pairing_code, option, MIN(strlen(option), 8));
```

Comment:
I can log an error. 

[NIST 800-73-4 Part 1](https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-73-4.pdf) "5.1.3 Pairing Code"
"If implemented, the pairing code shall consist of eight decimal digits and it shall be generated at
random by the PIV Card Issuer." It may not be required, as specified in the pin policy.  It is almost a second pin and may be printed in the card or not. On the test cards, they are all 8 ASCI digit numbers. 

How to use it for login over a contactless interface is still an issue (how to enter it or prompt user for it or look it up), so there is still work to do in this area.

Reply:
> How to use it for login over a contactless interface is still an issue (how to enter it or prompt user for it or look it up), so there is still work to do in this area.

User interaction for setting up an initial SM channel is easy to choose from a technical point of view, but it is one of the worst choice when it comes to UX. You can't make chocolate from shit, so I would not waste too many thoughts on that.

Well, actually I already did :wink: when we wanted to secure the contact-less sc-hsm... and we decided for a different solution. We used a static and signed ECDH key on the card for a key agreement (OpenSC chooses an ephemeral keypair). This way you only need to distribute the (public) root certificate of the signer instead of handling additional user input. Anyway, I also thought about an additional code. I think your best option is to request the code in a dedicated dialog, as done in DNIe's consent mechanism and to then cache the verified code inside OpenSC's on-disk file cache. But with respect to the limited use, you should stick to the configuration value in `opensc.conf` for now.

Reply:
I agree is very limited and not worth the effort at this time. The requirement to use a paring code is optional and set by the card issuer and used to setup Virtual Contact Interface over a contactless interface.  So cards with required pairing code used for login over a contactless reader could be quite small.